In [29]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import poisson
data = pd.read_csv('World Football Results 2018 to 2022 - data.csv')

In [ ]:
data.info()

In [10]:
data[data['tournament']=='Southeast Asian Championship']

year         date                    tournament       h_team  \
708   2018   November 8  Southeast Asian Championship     Cambodia   
709   2018   November 8  Southeast Asian Championship         Laos   
710   2018   November 9  Southeast Asian Championship    Singapore   
711   2018   November 9  Southeast Asian Championship     Thailand   
714   2018  November 12  Southeast Asian Championship     Malaysia   
715   2018  November 12  Southeast Asian Championship      Myanmar   
717   2018  November 13  Southeast Asian Championship    Indonesia   
718   2018  November 13  Southeast Asian Championship  Philippines   
752   2018  November 16  Southeast Asian Championship         Laos   
764   2018  November 16  Southeast Asian Championship      Vietnam   
781   2018  November 17  Southeast Asian Championship  Philippines   
786   2018  November 17  Southeast Asian Championship     Thailand   
832   2018  November 20  Southeast Asian Championship     Cambodia   
847   2018  November 20  Southeast Asian Championship      Myanmar   
860   2018  November 21  Southeast Asian Championship  Philippines   
861   2018  November 21  Southeast Asian Championship    Singapore   
862   2018  November 24  Southeast Asian Championship     Malaysia   
863   2018  November 24  Southeast Asian Championship      Vietnam   
864   2018  November 25  Southeast Asian Championship    Indonesia   
865   2018  November 25  Southeast Asian Championship     Thailand   
866   2018   December 1  Southeast Asian Championship     Malaysia   
867   2018   December 2  Southeast Asian Championship  Philippines   
868   2018   December 5  Southeast Asian Championship     Thailand   
869   2018   December 6  Southeast Asian Championship      Vietnam   
870   2018  December 11  Southeast Asian Championship     Malaysia   
872   2018  December 15  Southeast Asian Championship      Vietnam   
3115  2022    January 1  Southeast Asian Championship     Thailand   
4085  2022  December 20  Southeast Asian Championship     Thailand   
4086  2022  December 20  Southeast Asian Championship     Cambodia   
4087  2022  December 21  Southeast Asian Championship      Myanmar   
4088  2022  December 21  Southeast Asian Championship         Laos   
4090  2022  December 23  Southeast Asian Championship  Philippines   
4091  2022  December 23  Southeast Asian Championship    Indonesia   
4092  2022  December 24  Southeast Asian Championship    Singapore   
4093  2022  December 24  Southeast Asian Championship     Malaysia   
4094  2022  December 26  Southeast Asian Championship     Thailand   
4095  2022  December 26  Southeast Asian Championship    Indonesia   

           a_team  h_elo  a_elo  h_score  a_score            location neutral  
708      Malaysia    796   1069        0        1         in Cambodia      No  
709       Vietnam    730   1285        0        3             in Laos      No  
710     Indonesia   1068   1238        1        0        in Singapore      No  
711    East Timor   1398    652        7        0         in Thailand      No  
714          Laos   1080    725        3        1         in Malaysia      No  
715      Cambodia   1057    785        4        1          in Myanmar      No  
717    East Timor   1214    651        3        1        in Indonesia      No  
718     Singapore   1186   1092        1        0  in the Philippines      No  
752       Myanmar    721   1064        1        3             in Laos      No  
764      Malaysia   1290   1084        2        0          in Vietnam      No  
781    East Timor   1196    650        3        2         in Malaysia     Yes  
786     Indonesia   1399   1215        4        2         in Thailand      No  
832          Laos    778    709        3        1         in Cambodia      No  
847       Vietnam   1076   1299        0        0          in Myanmar      No  
860      Thailand   1198   1409        1        1  in the Philippines      No  
861    East Timor   1082    648        6        1        in Singa

In [3]:
data = data.dropna(axis=1)

In [4]:
data.isnull().sum()

year          0
date          0
tournament    0
h_team        0
a_team        0
h_elo         0
a_elo         0
h_score       0
a_score       0
location      0
neutral       0
dtype: int64

In [ ]:
data

In [6]:
data_home = data[['h_team', 'h_score', 'a_score']]
data_away = data[['a_team', 'h_score', 'a_score']]

In [25]:
data_home = data_home.rename(columns={'h_team': 'Team', 'h_score': 'GoalsScored', 'a_score': 'GoalsConceded'})
data_away = data_away.rename(columns={'a_team': 'Team', 'h_score': 'GoalsConceded', 'a_score': 'GoalsScored'})

In [28]:
team_strength = pd.concat([data_home, data_away], ignore_index=True).groupby('Team').mean()
team_strength

GoalsScored  GoalsConceded
Team                                   
Afghanistan     0.947368       1.947368
Albania         1.066667       1.200000
Algeria         2.018868       0.849057
Andorra         0.488889       1.711111
Angola          1.205882       0.735294
...                  ...            ...
Vietnam         1.710526       1.026316
Wales           1.000000       1.098039
Yemen           0.454545       2.090909
Zambia          1.384615       1.442308
Zimbabwe        0.921053       1.184211

[227 rows x 2 columns]

In [36]:
def predict_point(home, away):
    if home in team_strength.index and away in team_strength.index:
        lamb_home = team_strength.at[home, 'GoalsScored'] * team_strength.at[away, 'GoalsConceded']
        lamb_away = team_strength.at[away, 'GoalsScored'] * team_strength.at[home, 'GoalsConceded']
        prob_home, prob_away, prob_draw = 0, 0 , 0
        for x in range(0,11): #number of goals home team
            for y in range(0,11): #number of goals away team
                p = poisson.pmf(x, lamb_home) * poisson.pmf(y, lamb_away)
                if x == y:
                    prob_draw += p
                elif x > y:
                    prob_home += p
                else:
                    prob_away += p
        points_home = 3 * prob_home + prob_draw
        points_away = 3 * prob_away + prob_draw
        return[(points_home, points_away), (prob_home*100, prob_draw*100, prob_away*100)]
    else:
        return [(0,0), (prob_home*100, prob_draw*100, prob_away*100)]

In [37]:
predict_point('Vietnam', 'Singapore')

[(2.1109892380372846, 0.7216473605554833),
 (64.80390629075727, 16.687204931456627, 18.492510374697236)]

Vietnam wins the match.
Result: Viet Nam: 2.11 pts, Singapore: 0.72 pts


Vietnam win probability: Viet Nam: 64.8%, Draw: 16.7%, Singapore: 18.5%